wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java

```diff
@@ -138,20 +134,35 @@ public final void encode(OutputStream stream, E value) throws IOException {
 
   /** Read an encoded message from {@code bytes}. */
   public final E decode(byte[] bytes) throws IOException {
+    return decode(bytes, ExtensionRegistry.NO_EXTENSIONS);
+  }
+
+  /** Read an encoded message from {@code bytes}. */
+  public final E decode(byte[] bytes, ExtensionRegistry extensionRegistry) throws IOException {
     checkNotNull(bytes, "bytes == null");
-    return decode(new Buffer().write(bytes));
+    return decode(new Buffer().write(bytes), extensionRegistry);
   }
 
   /** Read an encoded message from {@code stream}. */
   public final E decode(InputStream stream) throws IOException {
+    return decode(stream, ExtensionRegistry.NO_EXTENSIONS);
+  }
+
+  public final E decode(InputStream stream, ExtensionRegistry extensionRegistry)
+      throws IOException {
     checkNotNull(stream, "stream == null");
-    return decode(Okio.buffer(Okio.source(stream)));
+    return decode(Okio.buffer(Okio.source(stream)), extensionRegistry);
   }
 
   /** Read an encoded message from {@code source}. */
   public final E decode(BufferedSource source) throws IOException {
+    return decode(source, ExtensionRegistry.NO_EXTENSIONS);
+  }
+
+  public final E decode(BufferedSource source, ExtensionRegistry extensionRegistry)
```

Comment:
now there's way too many overloads. I can actually cheat a bit here and put a little state back, `WireAdapter.withExtensions()` but that operation would be transitive for calls on that instance.